Config V2.0

jwaterfield opened the pull request:
New version of echidna which uses config files to specify dimensions to allow users to define their own variables. Supersedes #68

jwaterfield commented:
Hey @ashleyrback  I still need to add documentation but the code should all be there if you wanted to start reviewing. Cheers.

jwaterfield commented:
@ashleyrback  just realised plot_root needs converting still to.

ashleyrback commented:
@jwaterfield, apologies for the delay, haven't had much in the way of a reliable internet connection. The code all looks really good - all the logic seems to make sense. There are lots of errors when I try to build the docs (from your branch) though. Most of them seem to be `ImportError`s relating to imports from `echidna.output.store` or `echidna.output.plot`.

Also the unittests currently fail with three errors:

```
$ python -m unittest discover echidna/test/
......E...E.....E
======================================================================
ERROR: test_get_chi_squared (test_chi_squared.TestChiSquared)
Tests get chi squared method
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/ashley/snoplus/software/echidna/echidna-pr/echidna/test/test_chi_squared.py", line 114, in test_get_chi_squared
    mc_spectrum)
  File "echidna/limit/chi_squared.py", line 90, in get_chi_squared
    if (kwargs.get("penalty_terms") is not None):
NameError: global name 'kwargs' is not defined

======================================================================
ERROR: test_store (unittest.loader.ModuleImportFailure)
----------------------------------------------------------------------
ImportError: Failed to import test module: test_store
Traceback (most recent call last):
  File "/usr/lib/python2.7/unittest/loader.py", line 254, in _find_tests
    module = self._get_module_from_name(name)
  File "/usr/lib/python2.7/unittest/loader.py", line 232, in _get_module_from_name
    __import__(name)
  File "/home/ashley/snoplus/software/echidna/echidna-pr/echidna/test/test_store.py", line 3, in <module>
    import echidna.output.store as store
  File "echidna/output/store.py", line 4, in <module>
    import echidna.limit.limit_setting as limit_setting
  File "echidna/limit/limit_setting.py", line 7, in <module>
    import echidna.output.plot_root as plot
  File "echidna/output/plot_root.py", line 2, in <module>
    from echindna.output import root_style
  File "/home/ashley/snoing/install/root-5.34.30/lib/ROOT.py", line 352, in _importhook
    return _orig_ihook( name, *args, **kwds )
ImportError: No module named echindna.output


======================================================================
ERROR: test_limit_setting (unittest.loader.ModuleImportFailure)
----------------------------------------------------------------------
ImportError: Failed to import test module: test_limit_setting
Traceback (most recent call last):
  File "/usr/lib/python2.7/unittest/loader.py", line 254, in _find_tests
    module = self._get_module_from_name(name)
  File "/usr/lib/python2.7/unittest/loader.py", line 232, in _get_module_from_name
    __import__(name)
  File "/home/ashley/snoing/install/root-5.34.30/lib/ROOT.py", line 352, in _importhook
    return _orig_ihook( name, *args, **kwds )
  File "/home/ashley/snoplus/software/echidna/echidna-pr/echidna/test/test_limit_setting.py", line 7, in <module>
    import echidna.limit.limit_setting as limit_setting
  File "/home/ashley/snoing/install/root-5.34.30/lib/ROOT.py", line 352, in _importhook
    return _orig_ihook( name, *args, **kwds )
  File "echidna/limit/limit_setting.py", line 7, in <module>
    import echidna.output.plot_root as plot
  File "/home/ashley/snoing/install/root-5.34.30/lib/ROOT.py", line 352, in _importhook
    return _orig_ihook( name, *args, **kwds )
  File "echidna/output/plot_root.py", line 2, in <module>
    from echindna.output import root_style
  File "/home/ashley/snoing/install/root-5.34.30/lib/ROOT.py", line 352, in _importhook
    return _orig_ihook( name, *args, **kwds )
ImportError: No module named echindna.output


----------------------------------------------------------------------
Ran 17 tests in 9.897s

FAILED (errors=3)
```

I'll try and keep an eye onthis.

ashleyrback commented:
@jwaterfield sorry didn't mean to close it, pressed the wrong button.

ashleyrback commented on echidna/core/dsextract.py:
```diff
@@ -0,0 +1,608 @@
+
+    Args:
+      dimension (str): to extract from a RAT DS/ntuple file.
+      \**kwargs (dict): to be passed to and checked by the extractor.
```
@jwaterfield not sure what the `\**` is doing here. `kwargs` is currently not displayed in docs. See screenshot.

![screenshot from 2015-09-07 12 33 42](https://cloud.githubusercontent.com/assets/1931666/9715908/f63e4312-555c-11e5-9217-b2db62ad7f65.png)

ashleyrback commented on echidna/core/dsextract.py:
```diff
@@ -0,0 +1,608 @@
+      IndexError: dimension is an unknown parameter
+
+    Retuns:
+      Extractor object.
```
Could you fix the Returns section too so it gives the return type and description please.

ashleyrback commented on echidna/core/dsextract.py:
```diff
@@ -0,0 +1,608 @@
+
+    Args:
+      name (str): of the dimension
+      fid_r (float): Fiducial radius. Applies a cut to remove
```
I'm not sure about the `fid_r` as a variable name. Was there a reason for choosing `fid_r` over something like `fv_radius` that we had before? You still use `av_radius` elsewhere. It's completely personal preference, but I just feel `fv_radius` is more explicit than `fid_r`.

ashleyrback commented on echidna/core/dsextract.py:
```diff
@@ -0,0 +1,608 @@
+
+
+class Radial3ExtractMC(Extractor):
+    ''' True :math:`(radius/av_radius)^3`radial extraction methods.
```
sphinx-build warning:
```
/home/ashley/snoplus/software/echidna/echidna-pr/echidna/core/dsextract.py:docstring of echidna.core.dsextract.Radial3ExtractMC:1: WARNING: Inline interpreted text or phrase reference start-string without end-string.
```
I think a space before `radial` should fix this

ashleyrback commented on echidna/core/spectra.py:
```diff
@@ -1,4 +1,226 @@
+
+        .. note::
+
+        Keyword arguments include:
```
sphinx-build error:
```
/home/ashley/snoplus/software/echidna/echidna-pr/echidna/core/spectra.py:docstring of echidna.core.spectra.SpectraParameter.set_par:6: ERROR: Content block expected for the "note" directive; none found.
```
Need to indent this line and then further indent the list below.

ashleyrback commented on echidna/core/spectra.py:
```diff
@@ -188,73 +438,111 @@ def scale(self, num_decays):
-          time_low (float): Optional new high bound of the time.
-
-        Notes:
+          \**kwargs (float): Named parameters to slice on; note that these
```
Ah ok I've realised that you were using the `\**` to try and display the `**` in `**kwargs`. This doesn't seem to be working very well - see screenshot.
![screenshot from 2015-09-07 13 49 14](https://cloud.githubusercontent.com/assets/1931666/9717367/1aacd800-556a-11e5-821f-5bd9cb7ad94c.png)
I think it would possibly be best to ignore the `**` - `kwargs` is the actual variable name anyway, `**` is the operator that unpacks the values from the dictionary (see http://stackoverflow.com/a/1769478).

ashleyrback commented on echidna/core/spectra.py:
```diff
@@ -188,73 +438,111 @@ def scale(self, num_decays):
+
+        .. note:
+
           The logic in this method is the same for each dimension, first
```
sphinx-build warning:
```
/home/ashley/snoplus/software/echidna/echidna-pr/echidna/core/spectra.py:docstring of echidna.core.spectra.Spectra.shrink:13: WARNING: Explicit markup ends without a blank line; unexpected unindent.
```
I think this whole block needs to be indented.

ashleyrback commented on echidna/limit/chi_squared.py:
```diff
@@ -79,16 +74,11 @@ def get_chi_squared(self, observed, expected, **kwargs):
-
-        .. note::
-          Keyword arguments include
+      penalty_terms (dict, optional): specify (for each penalty term)
```
sphinx-build warning:
```
/home/ashley/snoplus/software/echidna/echidna-pr/echidna/limit/chi_squared.py:docstring of echidna.limit.chi_squared.ChiSquared.get_chi_squared:8: WARNING: Block quote ends without a blank line; unexpected unindent.
```
probably caused by messed-up indenting here.

ashleyrback commented on echidna/output/plot_chi_squared.py:
```diff
@@ -86,49 +86,41 @@ def chi_squared_vs_signal(signal_config, converter=None, fig_num=1,
-        analyser object, created during limit setting. Can be used
-        during limit setting setting or can load an instance from
-        hdf5
+      syst_analyser (:class:`echidna.limit.limit_setting.SystAnalyser`):
```
sphinx-build warning:
```
/home/ashley/snoplus/software/echidna/echidna-pr/echidna/output/plot_chi_squared.py:docstring of echidna.output.plot_chi_squared.chi_squared_map:6: WARNING: Block quote ends without a blank line; unexpected unindented.
```
I think you have to have a lease one word after the colon otherwise the formatting isn't correct.

![screenshot from 2015-09-07 16 42 25](https://cloud.githubusercontent.com/assets/1931666/9719785/e42390ba-557f-11e5-8b45-c5f24f6028b1.png)

ashleyrback commented on echidna/output/plot_chi_squared.py:
```diff
@@ -219,27 +212,25 @@ def chi_squared_map(syst_analyser, fig_num=1, **kwargs):
-        during limit setting setting or can load an instance from
-        hdf5
-      fig_num (int): Fig number. When creating multiple plots in the
+      syst_analyser (:class:`echidna.limit.limit_setting.SystAnalyser`):
```
sphinx-build warning:
```
/home/ashley/snoplus/software/echidna/echidna-pr/echidna/output/plot_chi_squared.py:docstring of echidna.output.plot_chi_squared.penalty_vs_systematic:6: WARNING: Block quote ends without a blank line; unexpected unindent.
```
Same again here.

ashleyrback commented on echidna/output/plot_chi_squared.py:
```diff
@@ -252,33 +243,32 @@ def penalty_vs_systematic(syst_analyser, fig_num=1, **kwargs):
-        analyser object, created during limit setting. Can be used
-        during limit setting setting or can load an instance from
-        hdf5
+      syst_analyser (:class:`echidna.limit.limit_setting.SystAnalyser`):
```
and here!
```
/home/ashley/snoplus/software/echidna/echidna-pr/echidna/output/plot_chi_squared.py:docstring of echidna.output.plot_chi_squared.push_pull:11: WARNING: Block quote ends without a blank line; unexpected unindent.
```

ashleyrback commented on echidna/output/plot_chi_squared.py:
```diff
@@ -356,18 +349,16 @@ def push_pull(syst_analyser, fig=1, **kwargs):
-        analyser object, created during limit setting. Can be used
-        during limit setting setting or can load an instance from
-        hdf5
+      syst_analyser (:class:`echidna.limit.limit_setting.SystAnalyser`):
```
and here!
```
/home/ashley/snoplus/software/echidna/echidna-pr/echidna/output/plot_chi_squared.py:docstring of echidna.output.plot_chi_squared.turn_on:8: WARNING: Block quote ends without a blank line; unexpected unindent.
```

ashleyrback commented:
@jwaterfield, all unittests pass now. Still some problems building the docs which I've commented on. Also want to run some of the example scripts to see how these work with config files.

ashleyrback commented on echidna/core/fill_spectrum.py:
```diff
@@ -8,427 +8,312 @@
+      spectrum (:class:`echidna.core.spectra.Spectra`): The spectrum which is
+        being filled.
+      chain (ROOT.TChain): Chain containing the events.
+      extractors (dict): Keys are the variable names and the keys are their
```
Typo: 2nd "keys" should be "values"

ashleyrback commented on echidna/core/fill_spectrum.py:
```diff
@@ -8,427 +8,296 @@
         new spectrum.
+      bipo (bool, optional): Applies the bipo cut if set to True.
+        Default is False.
+      \**kwargs (dict): Passed to and checked by the dsextractor.
```
@jwaterfield, A couple more docs issues I've noticed, starting with this one. There are still a couple of places with `\**` in front of `kwargs`. the `\` causes the whole line not to be included in the docs. I think it's fine just to write `kwargs` as the `**` is really an operator.

ashleyrback commented on echidna/core/fill_spectrum.py:
```diff
@@ -8,427 +8,296 @@
-def fill_mc_ntuple_spectrum(filename, T, spectrumname="", spectrum=None):
+def fill_from_root(filename, spectrum_name="", config=None, spectrum=None,
+                   bipo=False, **kwargs):
     """**Weights have been disabled.**
```
I think we can remove these lines now. There are also some references to weights in the lines below, which I'm not entirely sure about.

ashleyrback commented on echidna/core/dsextract.py:
```diff
@@ -0,0 +1,611 @@
+
+
+class Radial3ExtractMC(Extractor):
+    ''' True :math:`(radius/av_radius)^3` radial extraction methods.
```
I would prefer to use `outer_radius` as the variable name here instead of `av_radius`. The radius of the AV is a constant value 5997 mm. The purpose of this part was to bin with respect to some outer radius, which *is* the AV radius by default, but can be changed so you can consider other experiments e.g. KamLAND-Zen (inner balloon of radius 3080 mm). I know it's a subtle point but I think it would make it much easier to understand what is going on.

ashleyrback commented on echidna/scripts/dump_spectra_ntuple.py:
```diff
@@ -20,94 +19,82 @@
 
 
-def read_and_dump_ntuple(fname, half_life, spectrum_name, save_path):
+def read_and_dump_ntuple(fname, config_path, spectrum_name, save_path):
```
Can you add `**kwargs` to this method, which then get passed to `fill_spectrum.fill_from_ntuple`, so you can pass in keyword arguments here that are used in `fill_from_ntuple` and the extractors, e.g. `bipo`, 'fv_radius' and 'outer_radius'.

ashleyrback commented on echidna/scripts/dump_spectra_root.py:
```diff
@@ -20,92 +19,80 @@
 
 
-def read_and_dump_root(fname, half_life, spectrum_name, save_path):
+def read_and_dump_root(fname, config_path, spectrum_name, save_path):
```
again pass kwargs here

ashleyrback commented on echidna/scripts/multi_ntuple_spectrum.py:
```diff
@@ -21,56 +21,40 @@
 
 
-def create_combined_ntuple_spectrum(data_path, half_life,
+def create_combined_ntuple_spectrum(data_path, config_path,
```
and here.

jwaterfield commented:
Thats the way the bipo cuts work currently. The default is not to have the bipo cuts i.e. args.bipo is Fasle by default (done explicitly a few lines below L110). To switch them on a user would need to have the command --bipo in the command line. --no-bipo does the same as not including --bipo in the command line but I added it so a user can be explicit with their commands if they wanted to. I've added a note in the help of argparse to make it clearer to the user that bipo cuts are not applied by default. See next commit

ashleyrback commented:
@jwaterfield can you let me know when you've pushed the `args.` changes and fixed the error message. Cheers

ashleyrback commented:
Sorry just noticed you fixed the `args.`, can you just let me know when you've fixed the unexpected keyword argument error. Cheers

ashleyrback commented:
Cheers, looks like it's working now. I'll just finish re-creating the Majoron spectra and testing my the Majoron limits script, then it should be good to merge.

ashleyrback commented on echidna/output/plot.py:
```diff
@@ -23,31 +31,33 @@ def plot_projection(spectra, dimension, fig_num=1, show_plot=True):
+    x = _produce_axis(spectra, dimension)
+    par = spectra.get_config().get_par(dimension)
+    width = par.get_width()
+    plt.xlabel("%s (%s)", (dimension, unit))
```
A couple of errors during plotting. On this line `unit` is undefined, should be `par.get_unit()` I think

ashleyrback replied:
Also missing replace `,` with `%` for string formatting.

ashleyrback commented on echidna/output/plot.py:
```diff
@@ -2,17 +2,25 @@
 import matplotlib.pyplot as plt
 import numpy
 
+import echidna.core.spectra as spectra
```
Why is spectra imported here? It is not actually used as far as I can see. Also:

    from mpl_toolkits.mplot3d import Axes3D

can now be removed as this seems to be unused now.

ashleyrback commented on echidna/core/spectra.py:
```diff
@@ -1,4 +1,226 @@
+        Returns:
+          string: Unit of the parameter
+        """
+        if self._name.split['_'][0] == "energy":
```
Assuming the`par.get_unit()` change, I then get an error on this line. Should have normal (not square)brackets around split delimiter `('_')`

ashleyrback commented on echidna/output/plot.py:
```diff
@@ -23,31 +31,33 @@ def plot_projection(spectra, dimension, fig_num=1, show_plot=True):
+    plt.xlabel("%s (%s)", (dimension, unit))
+    plt.ylabel("Count per %f %s bin" % (width, par.get_unit()))
     data = spectra.project(dimension)
     ax.bar(x, data, width=width)
```
`ax` here should be `axis`

ashleyrback commented on echidna/core/dsextract.py:
```diff
@@ -0,0 +1,621 @@
+        Returns:
+          float: True energy
+        '''
+        entry.mcEdep
```
@jwaterfield, found out why my Majoron spectra were coming out with blank histograms. Missing return statement here, so the extractor was extracting the correct energy, but then returning `None`.

Also there are a couple of pep coding errors in `fill_spectrum` that I just noticed whilst bug hunting, some undefined variables and unused imports, theses should be flagged up by running pep8 on the file. I'll point them out later.

ashleyrback commented on echidna/core/fill_spectrum.py:
```diff
@@ -8,427 +8,294 @@
 from ROOT import TChain
 import math
```
```
rat imported but unused
math imported but unused
```

ashleyrback commented on echidna/core/fill_spectrum.py:
```diff
@@ -8,427 +8,294 @@
+        if check_next and entry.evIndex < 1:
+            try:
+                print "Filling"
+                spectrum.fill(**fill_kwargs)
```
undefined name `fill_kwargs`

ashleyrback commented on echidna/core/fill_spectrum.py:
```diff
@@ -8,427 +8,294 @@
-
-    for ievent in range(0, dsreader.GetEntryCount()):
+    for entry in range(0, dsreader.GetEntryCount()):
         ds = dsreader.GetEntry(ievent)
```
undefined name `ievent` --> should be `entry` I think.

ashleyrback commented on echidna/core/fill_spectrum.py:
```diff
@@ -3,432 +3,299 @@
             raise ValueError("Name not set when creating new spectra.")
-        spectrum = spectra.Spectra(str(spectrumname), 10.*chain.GetEntries())
+        spectrum = spectra.Spectra(str(spectrum_name),
+                                   dsreader.GetEntryCount(),
```
One final thing on this module, dsreader is undefined here, you define it below on line 277.

ashleyrback commented:
`multi_ntuple_spectrum` is now working perfectly for my Majoron spectra :+1:

ashleyrback commented:
@jwaterfield thanks for making those changes. I'm just smearing my Majoron spectra to check the majoron limits script. Two are finished but three jobs still going. As soon as it's done I'll post the results and should be fine to merge.

ashleyrback commented on echidna/output/store.py:
```diff
@@ -69,26 +132,72 @@ def load(file_path):
+                               file_.attrs["num_decays"],
+                               spectra_config)
+        # The following may not be present in old, old hdf5s
+        if file_.attrs["raw_events"]:
```
I tested this functionality as well, when comparing with old spectra. First this is really cool! But I've found it does still raise `KeyError` when these attributes don't exist in the HDF5. You could change it to something like this:

```
        try:
            spec._raw_events = file_.attrs["raw_events"]
        except KeyError as detail:  # unable to locate attribute, skip
            print detail
            print " --> skipping!"
        try:
            spec._style = file_.attrs["style"]
        except KeyError as detail:  # unable to locate attribute, skip
            print detail
            print " --> skipping!"
        try:
            spec._rois = file_.attrs["rois"]
        except KeyError as detail:  # unable to locate attribute, skip
            print detail
            print " --> skipping!"
```

Which I admit is not as nice, but I can't really think how else to do it.

jwaterfield replied:
Not sure whether this should be added or not? When this is merged then all spectra produced will have all the attributes in store.py. If they dont then that means something has gone wrong with storing them. The only spectra that require this change are the ones that were produced mid way through the pull request so shouldnt be used really anyway. I think if you need to use the spectra produced midway through the PR then these changes should just be held on a local branch. 

I agree something more clever needs to be done when future attrs gets added in later versions of echidna. Maybe a lookup between the version of echidna used to produce the spectra and the appropriate attrs for that version? I just think the try and except statements could just lead to masking bugs

This wouldnt need to be implemented until we need more attrs stored so maybe not a priority for this PR?

ashleyrback replied:
I agree. The only reason this came up was when I wanted to compare new spectra, produced with this PR and old spectra, i.e. produced before the config class was implemented. 

I agree that this should not be needed/used when the PR is merged. My main point was that the logic used doesn't seem to work if the attribute isn't stored in the hdf5, The if statements here raise `KeyError` instead of evaluating to `False` if not present. Maybe the entire method should not be included in the PR and only stored locally for "expert" operations.

The idea about a lookup table for attrs in future releases sounds good, but like you said we don't need to worry about that for this PR.

ashleyrback commented:
Since it is a major change, here is the verification document for this PR [pr_76_verification.pdf](https://github.com/snoplusuk/echidna/files/7888/pr_76_verification.pdf). As stated in the document, happy to merge, based on the following conditions:

* An updated version of the document will be uploaded after the spectra have be re-smeared and the analysis re-done with the re-smeared spectra
* An issue will be opened to check the scaling logic in `zero_nu_limit` (see #81)
* An issue will be opened to check the scaling logic in `Xe136_majoron_limits` (see #82)

ashleyrback commented:
Merging!
